code/streaming/Hacl.Streaming.Functor.fst

```diff
@@ -838,6 +867,155 @@ let modifies_footprint' c i p data len h0 h1 =
   if c.km = Runtime then
     c.key.frame_invariant #i (B.loc_addr_of_buffer p) k0 h0 h1
 
+/// Small helper which we use to factor out functional correctness proof obligations.
+/// It states that a state was correctly updated to process some data between two
+/// memory snapshots (the key is the same, the seen data was updated, etc.).
+///
+/// Note that we don't specify anything about the buffer and the block_state: they
+/// might have been updated differently depending on the case.
+unfold val state_is_updated :
+  #index:Type0 ->
+  (c: block index) ->
+  i:G.erased index -> (
+  let i = G.reveal i in
+  t:Type0 { t == c.state.s i } ->
+  t':Type0 { t' == optional_key i c.km c.key } ->
+  s:state c i t t' ->
+  data: B.buffer uint8 ->
+  len: UInt32.t ->
+  h0: HS.mem ->
+  h1: HS.mem ->
+  Type0)
+
+#push-options "--z3cliopt smt.arith.nl=false"
+let state_is_updated #index c i t t' p data len h0 h1 =
+  // Functional conditions about the memory updates
+  let s0 = B.get h0 p 0 in
+  let s1 = B.get h1 p 0 in
+  let State block_state0 buf0 total_len0 seen0 key0 = s0 in
+  let State block_state1 buf1 total_len1 seen1 key1 = s1 in
+
+  block_state1 == block_state0 /\
+  buf1 == buf0 /\
+  U64.v total_len1 == U64.v total_len0 + U32.v len /\
+  key1 == key0 /\
+
+  seen1 `S.equal` (seen0 `S.append` (B.as_seq h0 data)) /\
+
+  optional_reveal #index #i #c.km #c.key h1 key1 == optional_reveal #index #i #c.km #c.key h0 key0
+#pop-options
+
+/// The functional part of the invariant
```

Comment:
As far as I can see, this is a bit in-between the functional part of the invariant (the last two lines in invariant_s) and a small proof helper. Notice in particular that it takes two memories instead of just one. @sonmarcho would have a more accurate description.

Reply:
let's ask @sonmarcho to maybe write a comment as followup or something

Reply:
Yes indeed, @R1kM 's description is accurate.
The issue comes from the need to isolate some of the functional proof obligations, because some lemmas saturate. The most important part is what is at line 903, which holds for all the `update` functions and is easy to prove. From there, the exact goal to prove is more difficult to get and varies with the `update` function. I consequently use this small helper to factorize the preconditions of lemmas like `update_small_functional_correctness` or `update_empty_or_full_functional_correctness`.
Also note that I think we can cleanup the proofs a bit more: some lemma statements are too verbose (`update_small_functional_correctness` for instance) and some proofs can be factored out even more (`mk_finish`). I initially intended to do so but finally lacked time.